server/service/microsoft_mdm.go

```diff
@@ -0,0 +1,527 @@
+				Expires: getUtcTime(secWindowEndTimeMin),   // minutes from now
+			},
+		}
+		body.RequestSecurityTokenResponseCollection = &msg
+
+	case mdm_types.SoapFault:
+		// Setting the target action
+		if msg.OriginalMessageType == mdm_types.MDEDiscovery {
+			action = urlDiscovery
+		} else if msg.OriginalMessageType == mdm_types.MDEPolicy {
+			action = urlPolicy
+		} else if msg.OriginalMessageType == mdm_types.MDEEnrollment {
+			action = urlEnroll
+		} else {
+			action = urlDiag
+		}
+		uuid := uuid.New().String()
+		activityID = &mdm_types.ActivityId{
+			Content:       uuid,
+			CorrelationId: uuid,
+			XmlNS:         urlDiag,
+		}
+		body.SoapFault = &msg
+
+	default:
+		return SoapResponse{}, errors.New("mdm response message not supported")
+	}
+
+	// Return the SoapRequest type with the appropriate fields set
+	return SoapResponse{
+		XmlNSS: urlNSS,
+		XmlNSA: urlNSA,
+		XmlNSU: headerXsu,
+		Header: mdm_types.ResponseHeader{
+			Action: mdm_types.Action{
+				Content:        action,
+				MustUnderstand: MUValue,
+			},
+			RelatesTo:  relatesTo,
+			ActivityId: activityID,
+			Security:   security,
+		},
+		Body: body,
+	}, nil
+}
+
+// mdmMicrosoftDiscoveryEndpoint is the response struct for the GetDiscovery endpoint
+func mdmMicrosoftDiscoveryEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (errorer, error) {
+	req := request.(*SoapRequest)
+
+	// Checking first if Discovery message is valid and returning error if this is not the case
+	if err := req.IsValidDiscoveryMsg(); err != nil {
+		return getSoapResponseFault(req.Header.MessageID, mdm.SoapErrorMessageFormat, mdm_types.MDEDiscovery, err), nil
+	}
+
+	// Getting the DiscoveryResponse message
+	discoveryMessage, err := svc.GetMDMMicrosoftDiscoveryResponse(ctx)
+	if err != nil {
+		return getSoapResponseFault(req.Header.MessageID, mdm.SoapErrorMessageFormat, mdm_types.MDEDiscovery, err), nil
+	}
+
+	// Embedding the DiscoveryResponse message inside of a SoapResponse
+	response, err := NewSoapResponse(discoveryMessage, req.Header.MessageID)
+	if err != nil {
+		return getSoapResponseFault(req.Header.MessageID, mdm.SoapErrorMessageFormat, mdm_types.MDEDiscovery, err), nil
+	}
+
+	return response, nil
+}
+
+// GetMDMMicrosoftDiscoveryResponse returns a valid DiscoveryResponse message
+func (svc *Service) GetMDMMicrosoftDiscoveryResponse(ctx context.Context) (*fleet.DiscoverResponse, error) {
+	// skipauth: This endpoint does not use authentication
+	svc.authz.SkipAuthorization(ctx)
+
+	// Getting the app config
+	appCfg, err := svc.ds.AppConfig(ctx)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err)
+	}
+
+	// We can now craft the crafting DiscoveryResponse message
+	urlMDMServerDomain := appCfg.ServerSettings.ServerURL
+	urlDiscoveryEndpoint := urlMDMServerDomain + mdm.MDE2DiscoveryPath
+	urlPolicyEndpoint := urlMDMServerDomain + mdm.MDE2PolicyPath
+	urlEnrollEndpoint := urlMDMServerDomain + mdm.MDE2EnrollPath
```

Comment:
We generally avoid string-concatenating the base server URL with paths, and use path parsing instead to ensure it handles any trailing slash or other cases in the server URL (you adapt / use something like this when my PR merges: https://github.com/fleetdm/fleet/pull/12340/files#diff-2f8cc9b281dfe46edd1498300909a89ad5439eecee282f202c11102863c97b5bR10).

Reply:
Thanks! I've updated the code to use this